network.s.dfy

```diff
@@ -54,19 +59,26 @@ module Network {
   // allow it to be delivered over and over.
   // (We don't have packet headers, so duplication, besides being realistic,
   // also doubles as how multiple parties can hear the message.)
-  datatype Variables<Message> = Variables(sentMsgs:set<Message>)
+  datatype Variables<Payload> = Variables(correctlySignedMsgs:set<Message<Payload>>,
+                                                   sentMsgs:set<Message<Payload>>)
 
-  predicate Init<Message>(c: Constants, v: Variables<Message>)
+  predicate Init(c: Constants, v: Variables)
   {
     && v.sentMsgs == {}
+    && v.correctlySignedMsgs == {}
   }
 
-  predicate Next<Message>(c: Constants, v: Variables, v': Variables, msgOps: MessageOps)
+  predicate Next(c: Constants, v: Variables, v': Variables, msgOps: MessageOps, sender:HostId)
   {
     // Only allow receipt of a message if we've seen it has been sent.
     && (msgOps.recv.Some? ==> msgOps.recv.value in v.sentMsgs)
     // Record the sent message, if there was one.
-    && v'.sentMsgs ==
-      v.sentMsgs + if msgOps.send.None? then {} else { msgOps.send.value }
+    && (v'.sentMsgs ==
+        v.sentMsgs + if msgOps.send.Some? then { msgOps.send.value } else {})
+    && (v'.correctlySignedMsgs == v.correctlySignedMsgs 
+                                 + if msgOps.send.Some? && msgOps.send.value.sender == sender
+                                 then {msgOps.send.value} 
+                                 else {})
+    && msgOps.signatureChecks <= v.correctlySignedMsgs
```

Comment:
If you applied my comment above, this would read:
`&& msgOps.signedMsgs <= v.correctlySignedMsgs`

Reply:
Yes, this is indeed the intention here. We are going to use this for messages that pack and carry other messages from different peer replicas. This way we can check whether the packed messages are really valid. If we preserve the hint to the user it will be `&& msgOps.signedMsgsToCheck <= v.correctlySignedMsgs`, WDYT?